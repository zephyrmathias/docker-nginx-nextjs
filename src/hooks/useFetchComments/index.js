import { useState, useCallback } from 'react'
import axios from 'axios'

function useFetchComments() {
  const [comments, setComments] = useState([])
  const [error, setError] = useState(null)

  const fetchComments = useCallback(async (postId) => {
    try {
      setError(null)
      const { data } = await axios.get(
        `https://jsonplaceholder.typicode.com/posts/${postId}/comments`,
      )
      setComments(data)
    } catch (err) {
      setError(err)
    }
  }, [])

  return { comments, error, fetchComments }
}

export default useFetchComments
