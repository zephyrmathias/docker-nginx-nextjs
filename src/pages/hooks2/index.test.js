import React from 'react'
import { shallow } from 'enzyme'
import {
  render,
  waitForDomChange,
  act,
  getByText,
  fireEvent,
} from '@testing-library/react'
import CustomHooks2 from './index'
import useFetchComments from '../../hooks/useFetchComments'

jest.mock('../../hooks/useFetchComments')

test('should match with snapshot when it is loading', async () => {
  useFetchComments.mockImplementation(() => ({
    comments: [],
    error: null,
    fetchComments: jest.fn(),
  }))

  // use shallow so it does not trigger useEffect
  const component = shallow(<CustomHooks2 />)
  expect(component).toMatchSnapshot()
})

test('should match with snapshot when there is no data', async () => {
  const mockFn = {
    fetchComments: jest.fn(),
  }

  useFetchComments.mockImplementation(() => ({
    comments: [],
    error: null,
    fetchComments: mockFn.fetchComments,
  }))

  const fetchComments = jest.spyOn(mockFn, 'fetchComments')

  const { asFragment } = render(<CustomHooks2 />)

  await waitForDomChange()

  expect(asFragment()).toMatchSnapshot()
  expect(fetchComments).toHaveBeenNthCalledWith(1, 1) // to test useEffect
})

test('should match with snapshot when it contains a lits of comments', async () => {
  const mockFn = {
    fetchComments: jest.fn(),
  }

  useFetchComments.mockImplementation(() => ({
    comments: [
      {
        postId: 1,
        id: 1,
        name: 'name',
        email: 'email@email.com',
        body: 'body',
      },
    ],
    error: null,
    fetchComments: mockFn.fetchComments,
  }))

  const fetchComments = jest.spyOn(mockFn, 'fetchComments')

  const { asFragment } = render(<CustomHooks2 />)

  await waitForDomChange()

  expect(asFragment()).toMatchSnapshot()
  expect(fetchComments).toHaveBeenNthCalledWith(1, 1)
})

test('should match with snapshot when it contains error', async () => {
  const mockFn = {
    fetchComments: jest.fn(),
  }

  useFetchComments.mockImplementation(() => ({
    comments: [],
    error: 'error',
    fetchComments: mockFn.fetchComments,
  }))

  const fetchComments = jest.spyOn(mockFn, 'fetchComments')

  const { asFragment } = render(<CustomHooks2 />)

  await waitForDomChange()

  expect(asFragment()).toMatchSnapshot()
  expect(fetchComments).toHaveBeenNthCalledWith(1, 1)
})

test('should not set any value when component is unmounted', async () => {
  const mockFn = {
    fetchComments: jest.fn(),
  }

  useFetchComments.mockImplementation(() => ({
    comments: [],
    error: 'error',
    fetchComments: mockFn.fetchComments,
  }))

  const fetchComments = jest.spyOn(mockFn, 'fetchComments')

  const { asFragment, unmount } = render(<CustomHooks2 />)

  unmount()

  expect(asFragment()).toMatchSnapshot() // blank DOM since it's unmounted
  expect(fetchComments).toHaveBeenNthCalledWith(1, 1)
})

test('onClick button post id 1', async () => {
  useFetchComments.mockImplementation(() => ({
    comments: [
      {
        postId: 1,
        id: 1,
        name: 'name',
        email: 'email@email.com',
        body: 'body',
      },
    ],
    error: null,
    fetchComments: jest.fn(),
  }))

  const { asFragment, container } = render(<CustomHooks2 />)
  await waitForDomChange()
  const button = getByText(container, 'POST ID 1')
  fireEvent.click(button)
  expect(asFragment()).toMatchSnapshot()
})

test('onClick button post id 2', async () => {
  useFetchComments.mockImplementation(() => ({
    comments: [
      {
        postId: 1,
        id: 1,
        name: 'name',
        email: 'email@email.com',
        body: 'body',
      },
    ],
    error: null,
    fetchComments: jest.fn(),
  }))

  await act(async () => {
    const { asFragment, container } = render(<CustomHooks2 />)
    await waitForDomChange()
    const firstRender = asFragment()
    const button = getByText(container, 'POST ID 2')
    fireEvent.click(button)
    expect(asFragment()).toMatchDiffSnapshot(firstRender)
  })
})
