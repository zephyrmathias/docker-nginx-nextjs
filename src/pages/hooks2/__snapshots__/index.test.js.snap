// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onClick button post id 1 1`] = `
<DocumentFragment>
  <div>
    <h4>
      Example of how to use custom hooks (fetch) within useEffect (didmount and didupdate)
    </h4>
    <div>
      <button
        type="submit"
      >
        POST ID 1
      </button>
      <button
        type="submit"
      >
        POST ID 2
      </button>
    </div>
    <div>
      <h4>
        POST ID = 1
      </h4>
      <div>
        <div>
          name
        </div>
        <div>
          email@email.com
        </div>
        <div>
          body
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`onClick button post id 2 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -15,11 +15,11 @@
          POST ID 2
        </button>
      </div>
      <div>
        <h4>
-         POST ID = 2
+         POST ID = 1
        </h4>
        <div>
          <div>
            name
          </div>
`;

exports[`should match with snapshot when it contains a lits of comments 1`] = `
<DocumentFragment>
  <div>
    <h4>
      Example of how to use custom hooks (fetch) within useEffect (didmount and didupdate)
    </h4>
    <div>
      <button
        type="submit"
      >
        POST ID 1
      </button>
      <button
        type="submit"
      >
        POST ID 2
      </button>
    </div>
    <div>
      <h4>
        POST ID = 1
      </h4>
      <div>
        <div>
          name
        </div>
        <div>
          email@email.com
        </div>
        <div>
          body
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match with snapshot when it contains error 1`] = `
<DocumentFragment>
  <div>
    Error
  </div>
</DocumentFragment>
`;

exports[`should match with snapshot when it is loading 1`] = `
<div>
  Loading...
</div>
`;

exports[`should match with snapshot when there is no data 1`] = `
<DocumentFragment>
  <div>
    No data.
  </div>
</DocumentFragment>
`;

exports[`should not set any value when component is unmounted 1`] = `<DocumentFragment />`;
